// @linaria/webpack5-loader/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = webpack5Loader;

var _path = _interopRequireDefault(require("path"));

var _loaderUtils = _interopRequireDefault(require("loader-utils"));

var _enhancedResolve = _interopRequireDefault(require("enhanced-resolve"));

var _babelPreset = require("@linaria/babel-preset");

var _logger = require("@linaria/logger");

var _outputCssLoader = require("./outputCssLoader");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * This file contains a Webpack loader for Linaria.
 * It uses the transform.ts function to generate class names from source code,
 * returns transformed code without template literals and attaches generated source maps
 */
const outputCssLoader = require.resolve('./outputCssLoader');

function webpack5Loader(content, inputSourceMap) {
  var _this$_compilation, _webpackResolveOption, _result$sourceMap2;

  (0, _logger.debug)('loader', this.resourcePath);

  _babelPreset.EvalCache.clearForFile(this.resourcePath);

  const resolveOptionsDefaults = {
    conditionNames: ['require'],
    extensions: ['.js', '.jsx', '.ts', '.tsx', '.json']
  };
  const {
    sourceMap = undefined,
    preprocessor = undefined,
    extension = '.linaria.css',
    resolveOptions = {},
    ...rest
  } = this.getOptions() || {};
  const outputFileName = this.resourcePath.replace(/\.[^.]+$/, extension); // this._compilation is a deprecated API
  // However there seems to be no other way to access webpack's resolver
  // There is this.resolve, but it's asynchronous
  // Another option is to read the webpack.config.js, but it won't work for programmatic usage
  // This API is used by many loaders/plugins, so hope we're safe for a while

  const webpackResolveOptions = (_this$_compilation = this._compilation) === null || _this$_compilation === void 0 ? void 0 : _this$_compilation.options.resolve; // Resolved configuration contains empty list of extensions as a default value
  // https://github.com/callstack/linaria/issues/855

  if (((_webpackResolveOption = webpackResolveOptions.extensions) === null || _webpackResolveOption === void 0 ? void 0 : _webpackResolveOption.length) === 0) {
    delete webpackResolveOptions.extensions;
  } // Let's try to create a resolver with the webpack config


  let resolveSync = _enhancedResolve.default.create.sync({ ...resolveOptionsDefaults,
    ...(webpackResolveOptions !== null && webpackResolveOptions !== void 0 ? webpackResolveOptions : {}),
    ...resolveOptions,
    mainFields: ['main']
  });

  try {
    // Try to resolve the current file
    resolveSync(__dirname, __filename);
  } catch (e) {
    // Looks like one of the webpack plugins is async and the whole resolver became async
    (0, _logger.notify)('The default webpack configuration cannot be used because some of the plugins are asynchronous. All plugins have been ignored. Please override `resolveOptions.plugins` in the Linaria configuration.'); // Fallback to synchronous resolve

    resolveSync = _enhancedResolve.default.create.sync({ ...resolveOptionsDefaults,
      ...(webpackResolveOptions && {
        alias: webpackResolveOptions.alias,
        modules: webpackResolveOptions.modules
      } || {}),
      ...resolveOptions
    });
  }

  let result;
  const originalResolveFilename = _babelPreset.Module._resolveFilename;

  try {
    // Use webpack's resolution when evaluating modules
    _babelPreset.Module._resolveFilename = (id, {
      filename
    }) => {
      const result = resolveSync(_path.default.dirname(filename), id);

      if (!result) {
        // enhanced-resolve v4 throws a error when dependency is missed
        throw new Error('No result');
      }

      this.addDependency(result);
      return result;
    };

    result = (0, _babelPreset.transform)(content, {
      filename: _path.default.relative(process.cwd(), this.resourcePath),
      inputSourceMap: inputSourceMap !== null && inputSourceMap !== void 0 ? inputSourceMap : undefined,
      pluginOptions: rest,
      preprocessor
    });
  } finally {
    // Restore original behaviour
    _babelPreset.Module._resolveFilename = originalResolveFilename;
  }

  if (result.cssText) {
    var _result$dependencies, _result$sourceMap;

    let {
      cssText
    } = result;

    if (sourceMap) {
      cssText += `/*# sourceMappingURL=data:application/json;base64,${Buffer.from(result.cssSourceMapText || '').toString('base64')}*/`;
    }

    if ((_result$dependencies = result.dependencies) !== null && _result$dependencies !== void 0 && _result$dependencies.length) {
      result.dependencies.forEach(dep => {
        try {
          const f = resolveSync(_path.default.dirname(this.resourcePath), dep);
          this.addDependency(f);
        } catch (e) {
          // eslint-disable-next-line no-console
          console.warn(`[linaria] failed to add dependency for: ${dep}`, e);
        }
      });
    }

    (0, _outputCssLoader.addFile)(this.resourcePath, cssText);
    const request = `${outputFileName}!=!${outputCssLoader}!${this.resourcePath}`;

    const stringifiedRequest = _loaderUtils.default.stringifyRequest(this, request);

    this.callback(null, `${result.code}\n\nrequire(${stringifiedRequest});`, (_result$sourceMap = result.sourceMap) !== null && _result$sourceMap !== void 0 ? _result$sourceMap : undefined);
    return;
  }

  this.callback(null, result.code, (_result$sourceMap2 = result.sourceMap) !== null && _result$sourceMap2 !== void 0 ? _result$sourceMap2 : undefined);
}
